CIP-array-builtin/README.md

```diff
@@ -0,0 +1,286 @@
+
+We add the following builtin functions:
+
+1. `indexArray` of type `forall a . Integer -> Array a -> a`.
+    - It returns the element at the given index in the array, or fails with an error if the
+    index is outside the bounds of the array.
+    - It uses constant time and constant memory.
+2. `lengthOfArray` of type `forall a . Array a -> Integer`
+    - It returns the length of the array.
+    - It uses constant time and constant memory.
+3. `listToArray` of type `forall a . List a -> Array a`
+    - It converts the argument builtin list into a builtin array.
+    - It uses linear time and linear memory.
+
+### Binary serialisation and deserialisation
+
+As with all Plutus Core builtin types, arrays must have a fixed binary representation.
+
+For arrays we have chosen the following representation, based on the one currently
+implemented in the [flat][8] encoding for the [Haskell `Array` type][9].
+
+We define the result of encoding a Plutus Core array of type `forall a . Array a`
+and with length `n`, as follows:
+- Let:
+    - `Ea` be an encoding function for `a`, i.e. it is of type `a -> ByteString`.
+    - `Ei` be an encoding function for `Integer`
+    - `++` be bytestring concatenation
+    - `bin(i)` be the binary representation for any number `i`.
+- The first bits in the resulting sequence will be `Ei(0) ++ Ei(n-1)`, that is the
+encoding of the first index followed by the encoding of the last.
+- As binary, arrays are split into _blocks_ of `255` elements. For each block, the first
+bits in the block encoding can be decided as follows:
+    - there will be `n div 255` blocks where the next bits are `bin(255)`
+    - and, if `n mod 255 > 0`, one final block where the next bits are `bin(n mod 255)`. 
+- Inside each block, the next bits come from encoding all the elements pertaining to
+the block:
+    - blocks are indexed starting with `0`;
+    - `indexArray` is a function with the same behaviour as the proposed `indexArray` builtin,
+    since it is fixed, we omit the array argument for brevity;
+    - for the block with size `bs` at index `bn`, the bit sequence will be:
+    `Ea(index(bn*255)) ++ ... ++  Ea(index(bn*255 + bs - 1))`.
+- After all blocks have been encoded, the resulting binary sequence is prepended with exactly
+`8` bits of value `0`. 
+
+***Remark***: The decoding of arrays is, of course, the inverse of the encoding function.
+
+<details>
+    <summary>
+    Expand to see examples of applying the above encoding strategy to arrays of different element types.
+    </summary>
+
+    ------------
+    -- Example 1
+    >>> encode [1] as an array:
+    00000000 00000000 00000001 00000010 00000000
+
+    -- the first and last indices are 0, the number of elements is 1
+    -- 00000010 is the encoding of 1 of type Integer
+    -- the bytestring ends with 8 bits of value 0
+
+    ------------
+    -- Example 2
+    >>> encode [1, 1, 1] as an array:
+    00000000 00000100 00000011 00000010 00000010 00000010 00000000
```

Comment:
What is the 3 for?

Reply:
It's the number of elements in the block.

Reply:
> It's the number of elements in the block.
Oh right, I was confused because I thought it was part of the header for the whole thing.   If I may make a final suggstion, showing the encoding of something like `[11,22,33]` might make things a bit clearer since `[1,1,1]` doesn't convey a lot of information (and there's also the annoyance that it uses the zigzag encoding for integers so everything's shifted to the left by one bit).

Reply:
Sure, I'll add an example like that. Thanks!